### PR TITLE
SIP-112: Swap ETH for WETH in contract logic

### PR DESCRIPTION
**Pull request by liamzebedee**

### CODE REVIEW

**Comment by kaleb-keny** [comment]
Hey all good with the changes, thank you for incorporating them, although need to ask one thing, what if the user sends both wETH and ETH at the same time... I think then the logic would be to wrap the ETH sent added to the wETH received and then proceed with the calculations? Or does it revert because it kind of adds too much complexity to dealing with summing (meaning the cost of working with the contract would increase significantly)...
Also I believe i need to amend all the test cases to incorporate the case of receiving wETH and receiving ETH. As in the latter we don't have wrapping, while in the former ETH is wrapped... What do you think? I could help with that.

**Review by kaleb-keny — COMMENTED** [review]
Hey all good with the changes, thank you for incorporating them, although need to ask one thing, what if the user sends both wETH and ETH at the same time... I think then the logic would be to wrap the ETH sent added to the wETH received and then proceed with the calculations? Or does it revert because it kind of adds too much complexity to dealing with summing (meaning the cost of working with the contract would increase significantly)...
Also I believe i need to amend all the test cases to incorporate the case of receiving wETH and receiving ETH. As in the latter we don't have wrapping, while in the former ETH is wrapped... What do you think? I could help with that.

**Comment by liamzebedee** [comment]
> what if the user sends both wETH and ETH at the same time... I think then the logic would be to wrap the ETH sent added to the wETH received and then proceed with the calculations?

Good question, I didn't clarify this. How much demand is there for paying with both WETH+ETH? My guess is towards the low end, ready to be refuted here.

**Comment by liamzebedee** [comment]
>  Also I believe i need to amend all the test cases to incorporate the case of receiving wETH and receiving ETH. As in the latter we don't have wrapping, while in the former ETH is wrapped... What do you think? I could help with that.

If you'd like to write the test cases, that'd be great and I'll review. I think this change is mostly technical, rather than conceptual. So let me know how much detail you think is necessary. I'm still learning about SIP's and who reads em.

If it helps, here's a bit of a breakdown of the logic of accepting both:

```
mint(amount)
    if there is msg.value sent, then `amount` = msg.value, and the ETH has been received by the contract
    else, we assume amount is in WETH, and we transfer `amount` WETH to the contract

burn(amount, receiveETH)
    if receiveETH = true, then we allow withdrawals
```

**Comment by kaleb-keny** [comment]
> clarify

Yeah, i'd rather not have to deal with this complication 😅  but in case it does happen... of someone calling the contract and sending wETH and ETH at the same time, I'd rather not have any ETH locked in the contract without sETH being minted... As that would break the logic on burning sETH.... since more sETH can be burned later than is minted...

**Comment by kaleb-keny** [comment]
> Good question, I didn't clarify this. How much demand is there for paying with both WETH+ETH? My guess is towards the low end, ready to be refuted here.

Yeah, i'd rather not have to deal with this complication 😅 but in case it does happen... of someone calling the contract and sending wETH and ETH at the same time, I'd rather not have any ETH locked in the contract without sETH being minted... As that would break the logic on burning sETH.... since more sETH can be burned later than is minted...

**Comment by liamzebedee** [comment]
> in case it does happen... of someone calling the contract and sending wETH and ETH at the same time, I'd rather not have any ETH locked in the contract without sETH being minted

Yeah fully agree with what you're saying here 👍 

Clint and I had a bit of discussion today on #core about the design. I will probably design the contracts such that there is a base contract which only uses wETH, and then some utility code (possibly a [utility contract](https://github.com/Synthetixio/synthetix-utility)) which allows the user to transact with ETH. This will simplify the core interactions of what we're discussing, while allowing users to still interact with ETH. What are your thoughts / do you see any problems?

**Comment by kaleb-keny** [comment]
> > in case it does happen... of someone calling the contract and sending wETH and ETH at the same time, I'd rather not have any ETH locked in the contract without sETH being minted
> 
> Yeah fully agree with what you're saying here 👍
> 
> Clint and I had a bit of discussion today on #core about the design. I will probably design the contracts such that there is a base contract which only uses wETH, and then some utility code (possibly a [utility contract](https://github.com/Synthetixio/synthetix-utility)) which allows the user to transact with ETH. This will simplify the core interactions of what we're discussing, while allowing users to still interact with ETH. What are your thoughts / do you see any problems?

Yeah that would work I think, so just to confirm if I understood it right, the utility function is @payable and does not accept any tokens, what it can do basically wrap the ETH and sends it across to the main core wrapper function. The main core wrapper is no @payble so does not accept any ETH and can only take in wETH..... 
The minting/burning takes place on the core function... I think that would work just fine 👍🏼  As long we the core-contract can't hold any ETH by design... I think the implementation would be clean 😅

**Comment by kaleb-keny** [comment]
Can the utility be designed in a way that the user can't send wETH and send ETH simultaneously ?

**Comment by kaleb-keny** [comment]
Alright after some thinking, I think it's fine if wETH is sent to the contract  (without minting) and later burned ... as it'll just lower the debt on all minters... So i guess it's fine

File: sips/sip-112.md [hunk 1]
```diff
@@ -71,9 +73,18 @@ interface INativeEtherWrapper {
     function burn(uint amount) external;
 }
 ```
-
+#### L2 Specification
 `NativeEtherWrapper` is not deployed for L2, as using native ETH opcodes (`callvalue`, `balance`) in the OVM will revert.
 
+#### Key Bounds
+
+- The upper bound on the amount of  *Minting* is determined with a helper function, `capacity`, defined as `maxETH` less `wETH` locked in the contract. In case the user attempts to mint an amount greater than the upper bound, then `capacity` is minted and the residual is returned to the user (please refer to test cases for calculation specs).
```
Comment (liamzebedee) [inline sips/sip-112.md:81]: `capacity` is actually max(maxETH - weth, 0) if we're to be precise?

Comment (liamzebedee) [inline sips/sip-112.md:81]: nitpick - let's use consistent naming for wETH or WETH (I prefer the latter).

File: sips/sip-112.md [hunk 2]
```diff
@@ -71,9 +73,18 @@ interface INativeEtherWrapper {
...
+
+- The upper bound on the amount of *Burning* is computed as `wETH` locked in the contract multiplied `(1+burnFeeRate)` (please refer to test cases for calculation specs).
```
Comment (liamzebedee) [inline sips/sip-112.md:83]: 🙏 thanks for clarifying this one.

File: sips/sip-112.md (hunk 1 above)
Comment (kaleb-keny) [reply sips/sip-112.md]: sure, updated the full function

Comment (kaleb-keny) [reply sips/sip-112.md]: old habits die hard, 😅  all `wETH` changed to `WETH`

**Pull request merged**